### PR TITLE
symbol-clip: dynamic filter style spec changes

### PR DESCRIPTION
**Pull request by arindam1993**
This PR adds the necessary changes in the style-spec to account for dynamic filters ( filters with `["pitch"]` and `["distance-from-center"]`). The diff looks a bit weird because its a diff against the `symbol-clip` branch which is the branch the original prototype was built in using the `symbol-clip` layout property.

This support is only being added to symbol-layers so, to communicate that this PR adds some new root-level properties in the style-spec `v8.json`. 

- These are root level properties with keys formatted like `filter_${layerType}`. 
- The value for these new entries is an `expressionSpec` object, exactly like the expression spec objects used to declare expression compatibility for `paint`/`layout` properties.
- The updated style-spec validation code uses these new objects for filter validation, so there is a slight change in the formatting of error messages. https://github.com/mapbox/mapbox-gl-js/commit/450ca3f56d5a37b8dc4dd3e5e4ddc33429ee565e and https://github.com/mapbox/mapbox-gl-js/commit/c8d7ed539b78ae6683ec942b3bbd6cf8eaa8f673 show some of the new error messages.

### CODE REVIEW

File: debug/index.html [hunk 1]
```diff
@@ -27,33 +27,33 @@
...
-        ["<", ["pitch"], 60], false,
-        ["all", [">=", ["pitch"], 60], ["<", ["distance-from-center"], 3]], false,
-        true
-    ]
-    );
-
-    map.once('idle', () => {
-        map.panTo([-122.4199724, 37.7687162], {duration: 50000});
-    });
+    // map.setFilter('building-number-label',
```
Comment (avpeery) [inline debug/index.html:30]: Should there be a separate debug file for testing the commented out code instead of left commented out on the debug/index.html file?

Comment (rreusser) [reply debug/index.html:30]: Could be nice as a separate page with mapbox-gl-compare so you can see the difference this filter has.

File: src/style-spec/reference/v8.json [hunk 2]
```diff
@@ -652,7 +652,7 @@
     },
     "filter": {
       "type": "filter",
-      "doc": "A expression specifying conditions on source features. Only features that match the filter are displayed. Zoom expressions in filters are only evaluated at integer zoom levels. The `feature-state` expression is not supported in filter expressions."
+      "doc": "A expression specifying conditions on source features. Only features that match the filter are displayed. Zoom expressions in filters are only evaluated at integer zoom levels. The `feature-state` expression is not supported in filter expressions.  The `['pitch']` and `['distance-from-center']` expressions are supported only for filter expressions on the `symbol` layer."
```
Comment (rreusser) [inline src/style-spec/reference/v8.json:655]: Drive-by fix!

```suggestion
      "doc": "An expression specifying conditions on source features. Only features that match the filter are displayed. Zoom expressions in filters are only evaluated at integer zoom levels. The `feature-state` expression is not supported in filter expressions.  The `['pitch']` and `['distance-from-center']` expressions are supported only for filter expressions on the `symbol` layer."
```

Comment (rreusser) [reply src/style-spec/reference/v8.json:655]: Looks good, but it's getting to be quite a few very precise conditions in quick succession, which makes it a little challenging to parse. I think it's okay for now, but any more details will probably either require splitting this into a couple paragraphs or adding some more filler words or context to make it flow a little more smoothly.

File: src/style-spec/reference/v8.json [hunk 3]
```diff
@@ -2524,6 +2500,72 @@
     "value": "*",
     "doc": "A filter selects specific features from a layer."
   },
+  "filter_symbol": {
```
Comment (rreusser) [inline src/style-spec/reference/v8.json:2503]: Is there logic to whether these are snake_cased or camelCased? I definitely see both in the style-spec. Maybe it's that root properties are snake_cased and nested properties are camelCased?

File: src/style-spec/reference/v8.json [hunk 4]
```diff
@@ -2524,6 +2500,72 @@
...
+    "type": "boolean",
+    "doc": "Expression used to determine whether a symbol is being displayed or not. Symbols support `dynamic-filtering`, so this expression can account for the current state of the view by incorporating `[\"pitch\"]` and `[\"distance-from-center\"]` as part of the expression.",
```
Comment (rreusser) [inline src/style-spec/reference/v8.json:2505]: I think this description means to _define_ `dynamic-filtering`, but I read it like it was referencing an elsewhere-defined concept I should know about. I modified the wording a little bit, but I think the biggest change was "meaning" so that it reads like a self-contained definition.

```suggestion
    "doc": "Expression which determines whether or not to display a symbol. Symbols support `dynamic-filtering`, meaning this property can use the `[\"pitch\"]` and `[\"distance-from-center\"]` expressions to reference the current state of the view.",
```

File: src/style-spec/reference/v8.json [hunk 5]
```diff
@@ -2524,6 +2500,72 @@
...
+    "transition": false,
+    "property-type": "data-driven",
+    "expression": {
+      "interpolated": false,
+      "parameters": ["zoom", "feature", "pitch", "distance-from-center"]
+    }
+  },
+  "filter_fill": {
+    "type": "boolean",
+    "doc": "Expression used to determine whether a polygon is being displayed or not. Fill layer does NOT support `dynamic-filtering`, so this expression can NOT account for the current state of the view by incorporating `[\"pitch\"]` and `[\"distance-from-center\"]` as part of the expression.",
```
Comment (rreusser) [inline src/style-spec/reference/v8.json:2516]:
```suggestion
    "doc": "Expression which determines whether or not a polygon is displayed. Fill layer does *not* support `dynamic-filtering`, so this expression *cannot* use the `[\"pitch\"]` and `[\"distance-from-center\"]` expressions to reference the current state of the view.",
```

File: src/style-spec/reference/v8.json [hunk 6]
```diff
@@ -2524,6 +2500,72 @@
...
+    "transition": false,
+    "property-type": "data-driven",
+    "expression": {
+      "interpolated": false,
+      "parameters": ["zoom", "feature"]
+    }
+  },
+  "filter_line": {
+    "type": "boolean",
+    "doc": "Expression used to determine whether a LineString or a Polygon is being displayed or not. Line layer does NOT support `dynamic-filtering`, so this expression can NOT account for the current state of the view by incorporating `[\"pitch\"]` and `[\"distance-from-center\"]` as part of the expression.",
```
Comment (rreusser) [inline src/style-spec/reference/v8.json:2527]: Ditto, but I'll return to the rewording once we've nailed down the analogous wording for at least one property.

File: src/style-spec/reference/v8.json [hunk 7]
```diff
@@ -2524,6 +2500,72 @@
...
+    "transition": false,
+    "property-type": "data-driven",
+    "expression": {
+      "interpolated": false,
+      "parameters": ["zoom", "feature"]
+    }
+  },
+  "filter_fill-extrusion": {
+    "type": "boolean",
+    "doc": "Expression used to determine whether a polygon is being displayed or not. Fill-extrusion layer does NOT support `dynamic-filtering`, so this expression can NOT account for the current state of the view by incorporating `[\"pitch\"]` and `[\"distance-from-center\"]` as part of the expression.",
```
Comment (rreusser) [inline src/style-spec/reference/v8.json:2549]: Ditto above.

File: src/style-spec/reference/v8.json [hunk 8]
```diff
@@ -2524,6 +2500,72 @@
...
+    "transition": false,
+    "property-type": "data-driven",
+    "expression": {
+      "interpolated": false,
+      "parameters": ["zoom", "feature"]
+    }
+  },
+  "filter_heatmap": {
+    "type": "boolean",
+    "doc": "Expression used to determine whether a point is being displayed or not. Heatmap layer does NOT support `dynamic-filtering`, so this expression can NOT account for the current state of the view by incorporating `[\"pitch\"]` and `[\"distance-from-center\"]` as part of the expression.",
```
Comment (rreusser) [inline src/style-spec/reference/v8.json:2560]: Ditto above.

File: src/style-spec/reference/v8.json [hunk 9]
```diff
@@ -3183,6 +3225,24 @@
           }
         }
       },
+      "pitch": {
+        "doc": "Gets the current pitch in degrees. [\"pitch\"] may only be used in the `filter` expression for a `symbol` layer.",
```
Comment (rreusser) [inline src/style-spec/reference/v8.json:3229]:
```suggestion
        "doc": "Gets the current pitch in degrees. `[\"pitch\"]` may only be used in the `filter` expression for a `symbol` layer.",
```

File: src/style-spec/reference/v8.json [hunk 10]
```diff
@@ -3183,6 +3225,24 @@
...
+        "group": "Camera",
+        "sdk-support": {
+          "basic functionality": {
+            "js": "2.6.0"
+          }
+        }
+      },
+      "distance-from-center": {
+        "doc": "Returns the distance of a `symbol` instance from the center of the map. The distance is measured in pixels but normalized by the height of the map container. It measures 0 at the center, decreases towards the camera and increase away from the camera. For example, if the height of the map is 1000px, a value of -1 means 1000px away from the center towards the camera, and a value of 1 means a distance of 1000px away from the camera from the center.[\"distance-from-center\"] may only be used in the `filter` expression for a `symbol` layer.",
```
Comment (rreusser) [inline src/style-spec/reference/v8.json:3238]:
```suggestion
        "doc": "Returns the distance of a `symbol` instance from the center of the map. The distance is measured in pixels divided by the height of the map container. It measures 0 at the center, decreases towards the camera, and increase away from the camera. For example, if the height of the map is 1000px, a value of -1 means 1000px away from the center towards the camera, and a value of 1 means a distance of 1000px away from the camera from the center. `[\"distance-from-center\"]` may only be used in the `filter` expression for a `symbol` layer.",
```

**Review by rreusser — CHANGES_REQUESTED** [review]
Mainly some wording and doc touchups, but overall looking good. Excited for this feature. 🎉

File: src/style-spec/reference/v8.json (hunk 3 above)
Comment (arindam1993) [reply src/style-spec/reference/v8.json:2503]: In general its reverse snake case to denote nesting. like `layout_symbol` is used to denote the structure of `symbol.layout` object.
Hence the similar notation for this, `filter_symbol` denotes the structure of the `symbol.filter` object.

File: debug/index.html (hunk 1 above)
Comment (arindam1993) [reply debug/index.html]: I chose to get rid of this instead, I'll be adding a separate debug page in a followup PR that also adds render tests. Render-tests and the debug page will use the same geojson data so it should make debugging easier.

Not gonna go for gl-compare for this because it makes debugging harder because you can't easily tell which instance of the map is triggering a breakpoint.

File: test/unit/style-spec/fixture/filters-dynamic-distance.output-api-supported.json [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+  },
+  {
+    "message": "layers[4].filter: [\"distance-from-center\"] expression is not supported in a filter for a fill-extrusion layer with id: fill-extrusion-not-supported",
+    "line": 43
+  },
+  {
+    "message": "layers[5].filter: [\"distance-from-center\"] expression is not supported in a filter for a heatmap layer with id: heatmap-not-supported",
+    "line": 50
+  }
+]
```
Comment (avpeery) [inline test/unit/style-spec/fixture/filters-dynamic-distance.output-api-supported.json:22]: end with new line

File: src/style-spec/reference/v8.json (hunk 3 above)
Comment (rreusser) [reply src/style-spec/reference/v8.json:2503]: Thanks for the clarification!

File: test/unit/style-spec/fixture/filters-dynamic-distance.output-api-supported.json (hunk 11 above)
Comment (avpeery) [reply test/unit/style-spec/fixture/filters-dynamic-distance.output-api-supported.json:22]: Aside from adding new lines at the end of the tests, this lgtm!

File: src/style-spec/feature_filter/index.js [hunk 12]
```diff
@@ -53,75 +54,75 @@ function isExpressionFilter(filter: any) {
...
+
         return {
-            filter: (globalProperties: GlobalProperties, feature: Feature, canonical?: CanonicalTileID) => compiledStaticFilter.value.evaluate(globalProperties, feature, {}, canonical),
-            dynamicFilter: (globalProperties: GlobalProperties, feature: Feature, canonical?: CanonicalTileID, refLocation: ?MercatorCoordinate) => compiledDynamicFilter.value.evaluate(globalProperties, feature, {}, canonical, null, null, refLocation),
+            filter: filterFunc,
+            dynamicFilter: dynamicFilterFunc ? dynamicFilterFunc : undefined,
             needGeometry
         };
+    } else {
+        // This branch cannot happen but need to keep flow happy :(
```
Comment (SnailBones) [inline src/style-spec/feature_filter/index.js:124]: It seems like Flow is dictating the structure of this whole section at the expense of readability!
Would Flow be ok with a` if (!filterFunc) {return;}` so that you could unwrap the main return statement?
Or could you make `filterFunc` a `const` and use a try / catch?

File: src/style-spec/reference/v8.json [hunk 13]
```diff
@@ -652,7 +652,7 @@
     },
     "filter": {
       "type": "filter",
-      "doc": "A expression specifying conditions on source features. Only features that match the filter are displayed. Zoom expressions in filters are only evaluated at integer zoom levels. The `feature-state` expression is not supported in filter expressions."
+      "doc": "An expression specifying conditions on source features. Only features that match the filter are displayed. Zoom expressions in filters are only evaluated at integer zoom levels. The `feature-state` expression is not supported in filter expressions.  The `['pitch']` and `['distance-from-center']` expressions are supported only for filter expressions on the `symbol` layer."
```
Comment (SnailBones) [inline src/style-spec/reference/v8.json:655]:
```suggestion
      "doc": "An expression specifying conditions on source features. Only features that match the filter are displayed. [`zoom`](#zoom) expressions in filters are only evaluated at integer zoom levels. The [`feature-state`](#feature-state) expression is not supported in filter expressions.  The [`pitch'](#pitch) and [`distance-from-center'`](distance-from-center) expressions are supported only for filter expressions on the [`symbol`](/mapbox-gl-js/style-spec/layers/#symbol) layer."
```

File: src/style-spec/reference/v8.json [hunk 14]
```diff
@@ -2524,6 +2500,72 @@
     "value": "*",
     "doc": "A filter selects specific features from a layer."
   },
+  "filter_symbol": {
+    "type": "boolean",
+    "doc": "Expression which determines whether or not to display a symbol. Symbols support `dynamic-filtering`, meaning this expression can use the `[\"pitch\"]` and `[\"distance-from-center\"]` expressions to reference the current state of the view.",
```
Comment (SnailBones) [inline src/style-spec/reference/v8.json:2505]: Nit:
Is `dynamic-filtering` a string that users will need to type? If not, I think should just be written in plain text as dynamic filtering.

I'm noticing that we have an inconsistency in the style-spec for whether to write expressions as: `expression-name`,  `'expression-name'`, [\"expression-name\"], or `[\"expression-name\"]`. I'd love to hear from the docs team on this.

File: src/style-spec/reference/v8.json [hunk 15]
```diff
@@ -4713,9 +4755,6 @@
       "doc": "Defines a gradient with which to color a line feature. Can only be used with GeoJSON sources that specify `\"lineMetrics\": true`.",
       "transition": false,
       "requires": [
-        {
-          "!": "line-dasharray"
```
Comment (SnailBones) [inline src/style-spec/reference/v8.json:4717]: Good catch!

File: src/style-spec/reference/v8.json [hunk 16]
```diff
@@ -2524,6 +2500,72 @@
...
+    "transition": false,
+    "property-type": "data-driven",
+    "expression": {
+      "interpolated": false,
+      "parameters": ["zoom", "feature"]
+    }
+  },
+  "filter_heatmap": {
+    "type": "boolean",
+    "doc": "Expression used to determine whether a point is being displayed or not. Heatmap layer does NOT support `dynamic-filtering`, meaning this expression can NOT use the `[\"pitch\"]` and `[\"distance-from-center\"]` expressions to reference the current state of the view.",
```
Comment (SnailBones) [inline src/style-spec/reference/v8.json:2560]: Rephrase for consistency

File: src/style-spec/reference/v8.json [hunk 17]
```diff
@@ -3183,6 +3225,24 @@
           }
         }
       },
+      "pitch": {
+        "doc": "Gets the current pitch in degrees. `[\"pitch\"]` may only be used in the `filter` expression for a `symbol` layer.",
```
Comment (SnailBones) [inline src/style-spec/reference/v8.json:3229]: Nit: consider "Returns" instead of "Gets"

File: src/style-spec/validate/validate_expression.js [hunk 18]
```diff
@@ -41,3 +44,31 @@ export default function validateExpression(options: any): Array<ValidationError>
...
+    }
+
+    if (disallowedParameters.size === 0) {
+        return [];
+    }
+    const errors = [];
+
+    if (e instanceof CompoundExpression) {
+        if (disallowedParameters.has(e.name)) {
+            return [new ValidationError(options.key, options.value, `["${e.name}"] expression is not supported in a filter for a ${options.object.type} layer with id: ${options.object.id}`)];
```
Comment (SnailBones) [inline src/style-spec/validate/validate_expression.js:66]: Stylistic question here! Do we prefer to put the correct-use-case return in a nested value and error returns (or throws) at the end? I usually do it the other way around. In any case I'd like to be consistent!

This applies to line 25 too.

**Comment by SnailBones** [comment]
Great work @arindam1993! I have a few nits about code styling and documentation syntax.
A nice-to-have would be more links in the documentation.
I think that we should get some eyes from the docs team on this. Do you think it's best to do that in one go after it's merged into the `symbol-clip` branch?

File: src/style-spec/reference/v8.json (hunk 17 above)
Comment (arindam1993) [reply src/style-spec/reference/v8.json:3229]: This is consistent with the wording for `["zoom"]`, do you think we should change both?

Comment (SnailBones) [reply src/style-spec/reference/v8.json:3229]: Yes, I think "Returns" is slightly more clear.

File: src/style-spec/reference/v8.json (hunk 13 above)
Comment (arindam1993) [reply src/style-spec/reference/v8.json]: This is a good idea but i don't think we should do this manually, instead the docs codegen system should handle this, that way if the structure of the docs website changes it can account for it.

File: src/style-spec/feature_filter/index.js (hunk 12 above)
Comment (arindam1993) [reply src/style-spec/feature_filter/index.js:124]: Nope, flow cannot detect early returns statically.

Comment (arindam1993) [reply src/style-spec/feature_filter/index.js]: I did a force cast instead, looks slightly cleaner.

**Comment by arindam1993** [comment]
> Great work @arindam1993! I have a few nits about code styling and documentation syntax. A nice-to-have would be more links in the documentation. I think that we should get some eyes from the docs team on this. Do you think it's best to do that in one go after it's merged into the `symbol-clip` branch?

If we consistently name expressions, I think its something the docs-page-generation tooling should handle. That way it can account for changes to the website hierarchy without that having to be hardcoded in here.

File: src/style-spec/reference/v8.json (hunk 13 above)
Comment (SnailBones) [reply src/style-spec/reference/v8.json]: That sounds like a good idea to me!

File: src/style-spec/validate/validate_expression.js (hunk 18 above)
Comment (arindam1993) [reply src/style-spec/validate/validate_expression.js:66]: I don't think we have a hard rule. I think doing it a way that makes it easier to read a particular function is better. I personally like error handling/validation at the top and doing the meat of the logic after.

**Comment by SnailBones** [comment]
Just tested this with the docs and it seems to be breaking.
```
types.js:40 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'sort')
    at types.js:40
    at Array.map (<anonymous>)
    at Module.T1xj (types.js:37)
    at __webpack_require__ (bootstrap:84)
    at Module.flda (mapbox-gl-js-api~mapbox-gl-js-api-events~mapbox-gl-js-api-geography~mapbox-gl-js-api-handlers~mapbox~559181ac.chunk.js:6205)
    at __webpack_require__ (bootstrap:84)
    at Module.1wO5 (mapbox-gl-js-api~mapbox-gl-js-api-events~mapbox-gl-js-api-geography~mapbox-gl-js-api-handlers~mapbox~559181ac.chunk.js:694)
    at __webpack_require__ (bootstrap:84)
    at Object.ZwM2 (query-terrain-elevation.md:22)
    at __webpack_require__ (bootstrap:84)
```

File: src/style-spec/reference/v8.json [hunk 19]
```diff
@@ -3127,7 +3169,7 @@
         }
       },
       "properties": {
-        "doc": "Gets the feature properties object.  Note that in some cases, it may be more efficient to use [\"get\", \"property_name\"] directly.",
+        "doc": "Returns the feature properties object.  Note that in some cases, it may be more efficient to use [\"get\", \"property_name\"] directly.",
```
Comment (rreusser) [inline src/style-spec/reference/v8.json:3172]: Do we want code formatting, for consistency? Only a tiny nit, but just to make sure downstream docs are rendered correctly:

```suggestion
        "doc": "Returns the feature properties object.  Note that in some cases, it may be more efficient to use `[\"get\", \"property_name\"]` directly.",
```

File: src/style-spec/reference/v8.json [hunk 20]
```diff
@@ -3183,8 +3225,26 @@
...
+        "doc": "Returns the current pitch in degrees. `[\"pitch\"]` may only be used in the `filter` expression for a `symbol` layer.",
+        "group": "Camera",
+        "sdk-support": {
+          "basic functionality": {
+            "js": "2.6.0"
+          }
+        }
+      },
+      "distance-from-center": {
+        "doc": "Returns the distance of a `symbol` instance from the center of the map. The distance is measured in pixels divided by the height of the map container. It measures 0 at the center, decreases towards the camera and increase away from the camera. For example, if the height of the map is 1000px, a value of -1 means 1000px away from the center towards the camera, and a value of 1 means a distance of 1000px away from the camera from the center. [\"distance-from-center\"] may only be used in the `filter` expression for a `symbol` layer.",
```
Comment (rreusser) [inline src/style-spec/reference/v8.json:3238]: Code formatting nit:

```suggestion
        "doc": "Returns the distance of a `symbol` instance from the center of the map. The distance is measured in pixels divided by the height of the map container. It measures 0 at the center, decreases towards the camera and increase away from the camera. For example, if the height of the map is 1000px, a value of -1 means 1000px away from the center towards the camera, and a value of 1 means a distance of 1000px away from the camera from the center. `[\"distance-from-center\"]` may only be used in the `filter` expression for a `symbol` layer.",
```

**Review by rreusser — APPROVED** [review]
